osquery/core/database/rocksdb_migration.cpp

```diff
@@ -0,0 +1,281 @@
+  handle.options.error_if_exists = error_if_exists;
+  handle.path = path;
+  std::vector<std::string> column_families;
+  rocksdb::DB::ListColumnFamilies(handle.options, path, &column_families);
+
+  std::vector<rocksdb::ColumnFamilyDescriptor> descriptors;
+  for (const auto& column : column_families) {
+    descriptors.push_back(
+        rocksdb::ColumnFamilyDescriptor(column, handle.options));
+  }
+  std::vector<rocksdb::ColumnFamilyHandle*> column_family_handles;
+  rocksdb::DB* db = nullptr;
+  auto status = rocksdb::DB::Open(
+      handle.options, path, descriptors, &column_family_handles, &db);
+  if (status.IsInvalidArgument()) {
+    return createError(RocksdbMigrationError::InvalidArgument,
+                       status.ToString());
+  }
+  if (!status.ok()) {
+    return createError(RocksdbMigrationError::FailToOpen, status.ToString());
+  }
+  handle.db_handle = std::unique_ptr<rocksdb::DB>(db);
+  for (const auto& ptr : column_family_handles) {
+    handle.handles[ptr->GetName()] =
+        std::unique_ptr<rocksdb::ColumnFamilyHandle>(ptr);
+  }
+  return std::move(handle);
+}
+
+Expected<int, RocksdbMigrationError> RocksdbMigration::getVersion(
+    const DatabaseHandle& db) {
+  rocksdb::ReadOptions options;
+  options.verify_checksums = true;
+  auto handle_iter = db.handles.find("configurations");
+  if (handle_iter != db.handles.end()) {
+    // Try to get new version first
+    // Version stored as string value to help analyze db by tools
+    std::string version_str;
+
+    rocksdb::ColumnFamilyHandle* raw_handle = handle_iter->second.get();
+
+    auto status = db.db_handle->Get(options,
+                                    raw_handle,
+                                    std::string("database_schema_version"),
+                                    &version_str);
+    if (status.IsNotFound()) {
+      // Fallback to old version storage
+      auto default_family_handle_iter = db.handles.find("default");
+      if (default_family_handle_iter != db.handles.end()) {
+        status = db.db_handle->Get(options,
+                                   default_family_handle_iter->second.get(),
+                                   std::string("results_version"),
+                                   &version_str);
+      }
+    }
+    if (!status.ok()) {
+      return createError(RocksdbMigrationError::FailToGetVersion,
+                         status.ToString());
+    }
+    auto result = tryTo<int>(version_str);
+    if (result) {
+      return *result;
+    }
+    return createError(
+        RocksdbMigrationError::FailToGetVersion, "", result.takeError());
+  }
+  return createError(RocksdbMigrationError::FailToGetVersion,
+                     "Verion data is not found");
+}
+
+std::string RocksdbMigration::randomOutputPath() {
+  auto path = fs::path(OSQUERY_HOME);
+  boost::uuids::uuid uuid = boost::uuids::random_generator()();
+  path.append("migration");
+  path.append(boost::uuids::to_string(uuid));
+  return path.string();
```

Comment:
Would you like to use unique_path from boost instead of this function?
```c++
boost::filesystem::unique_path(OSQUERY_HOME "migration.%%%%-%%%%-%%%%");
```

Reply:
filesystem::unique_path is throwable and may fail, so I will stick for non failable function for now,  till we find out better to handle chained Expected. Since handling error for this function will produce a lot of unnecessary code